src/ledger.rs

```diff
@@ -1,18 +1,162 @@
+pub struct LedgerWindow {
+    index: File,
+    data: File,
+}
+
+// use a CONST because there's a cast, and we don't want "sizeof::<u64> as u64"...
+const SIZEOF_U64: u64 = size_of::<u64>() as u64;
+
+impl LedgerWindow {
+    // opens a Ledger in directory, provides "infinite" window
+    pub fn new(directory: &str) -> io::Result<Self> {
+        let directory = Path::new(&directory);
+
+        let index = File::open(directory.join("index"))?;
+        let data = File::open(directory.join("data"))?;
+
+        Ok(LedgerWindow { index, data })
+    }
+
+    pub fn get_entry(&mut self, index: u64) -> io::Result<Entry> {
+        fn u64_at(file: &mut File, at: u64) -> io::Result<u64> {
+            file.seek(SeekFrom::Start(at))?;
+            deserialize_from(file.take(SIZEOF_U64))
+                .map_err(|e| io::Error::new(ErrorKind::Other, e.to_string()))
+        }
+
+        let end_offset = u64_at(&mut self.index, index * SIZEOF_U64)?;
+
+        let start_offset = if index != 0 {
+            u64_at(&mut self.index, (index - 1) * SIZEOF_U64)?
+        } else {
+            0u64
+        };
+
+        fn entry_at(file: &mut File, at: u64, len: u64) -> io::Result<Entry> {
+            file.seek(SeekFrom::Start(at))?;
+
+            deserialize_from(file.take(len))
+                .map_err(|e| io::Error::new(ErrorKind::Other, e.to_string()))
+        }
+
+        entry_at(&mut self.data, start_offset, end_offset - start_offset)
+    }
+}
+
+#[derive(Debug)]
+pub struct LedgerWriter {
+    index: File,
+    data: File,
+}
+
+impl LedgerWriter {
+    // opens or creates a LedgerWriter in directory
+    pub fn new(directory: &str) -> io::Result<Self> {
+        let directory = Path::new(&directory);
+
+        create_dir_all(directory)?;
+
+        let index = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(directory.join("index"))?;
+
+        let data = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(directory.join("data"))?;
+
+        Ok(LedgerWriter { index, data })
+    }
+
+    fn write_entry(&mut self, entry: &Entry) -> io::Result<()> {
+        serialize_into(&mut self.data, &entry)
+            .map_err(|e| io::Error::new(ErrorKind::Other, e.to_string()))?;
+        self.data.flush()?;
+
+        let offset = self.data.seek(SeekFrom::Current(0))?;
+        serialize_into(&mut self.index, &offset)
+            .map_err(|e| io::Error::new(ErrorKind::Other, e.to_string()))?;
+        self.index.flush()
+    }
+
+    pub fn write_entries<I>(&mut self, entries: I) -> io::Result<()>
+    where
+        I: IntoIterator<Item = Entry>,
+    {
+        for entry in entries {
+            self.write_entry(&entry)?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct LedgerReader {
+    offset: u64, // next start_offset
+    index: File,
+    data: File,
+}
+
+impl Iterator for LedgerReader {
+    type Item = io::Result<Entry>;
```

Comment:
How about a `bincode::Result`?

Reply:
will raise an issue for this one, something like "give ledger its own errors" or "look into error_chain across the board for solana", or "make ledger able to return multiple types of errors" ?